### PR TITLE
Dev support instant id

### PR DESCRIPTION
**Pull request by chengzeyi**
```
python3 benchmarks/instant_id.py --repo ~/repos/InstantID/ --face-analysis-root ~/repos/face-analysis-root/
```

### CODE REVIEW

**Comment by chengzeyi** [comment]
```
Traceback (most recent call last):
  File "/home/cheng/repos/onediff/src/onediff/infer_compiler/with_oneflow_compile.py", line 172, in wrapper
    return func(self, *args, **kwargs)
  File "/home/cheng/repos/onediff/src/onediff/infer_compiler/utils/graph_management_utils.py", line 82, in wrapper
    ret = func(self, *args, **kwargs)
  File "/home/cheng/repos/onediff/src/onediff/infer_compiler/with_oneflow_compile.py", line 255, in __call__
    output = dpl_graph(*args, **kwargs)
  File "/home/cheng/repos/oneflow/python/oneflow/nn/graph/graph.py", line 292, in __call__
    return self._dynamic_input_graph_cache(*args, **kwargs)
  File "/home/cheng/repos/oneflow/python/oneflow/nn/graph/cache.py", line 116, in __call__
    return graph(*args, **kwargs)
  File "/home/cheng/repos/oneflow/python/oneflow/nn/graph/graph.py", line 295, in __call__
    self._compile(*args, **kwargs)
  File "/home/cheng/repos/oneflow/python/oneflow/nn/graph/graph.py", line 865, in _compile
    return self._compile_new(*args, **kwargs)
  File "/home/cheng/repos/oneflow/python/oneflow/nn/graph/graph.py", line 884, in _compile_new
    _, eager_outputs = self.build_graph(*args, **kwargs)
  File "/home/cheng/repos/oneflow/python/oneflow/nn/graph/graph.py", line 1428, in build_graph
    outputs = self.__build_graph(*args, **kwargs)
  File "/home/cheng/repos/oneflow/python/oneflow/nn/graph/graph.py", line 1576, in __build_graph
    outputs = self.build(*lazy_args, **lazy_kwargs)
  File "/home/cheng/repos/onediff/src/onediff/infer_compiler/with_oneflow_compile.py", line 322, in build
    return self.model(*args, **kwargs)
  File "/home/cheng/repos/oneflow/python/oneflow/nn/graph/proxy.py", line 188, in __call__
    result = self.__block_forward(*args, **kwargs)
  File "/home/cheng/repos/oneflow/python/oneflow/nn/graph/proxy.py", line 238, in __block_forward
    result = unbound_forward_of_module_instance(self, *args, **kwargs)
  File "/home/cheng/repos/onediff/src/infer_compiler_registry/register_diffusers/unet_2d_condition_oflow.py", line 358, in forward
    sample, res_samples = downsample_block(
  File "/home/cheng/repos/oneflow/python/oneflow/nn/graph/proxy.py", line 188, in __call__
    result = self.__block_forward(*args, **kwargs)
  File "/home/cheng/repos/oneflow/python/oneflow/nn/graph/proxy.py", line 238, in __block_forward
    result = unbound_forward_of_module_instance(self, *args, **kwargs)
  File "/home/cheng/pyvenv/oneflow/lib/python3.10/site-packages/diffusers/models/unet_2d_blocks.py", line 1160, in forward
    hidden_states = attn(
  File "/home/cheng/repos/oneflow/python/oneflow/nn/graph/proxy.py", line 188, in __call__
    result = self.__block_forward(*args, **kwargs)
  File "/home/cheng/repos/oneflow/python/oneflow/nn/graph/proxy.py", line 238, in __block_forward
    result = unbound_forward_of_module_instance(self, *args, **kwargs)
  File "/home/cheng/repos/onediff/src/infer_compiler_registry/register_diffusers/transformer_2d_oflow.py", line 1329, in forward
    hidden_states = block(
  File "/home/cheng/repos/oneflow/python/oneflow/nn/graph/proxy.py", line 188, in __call__
    result = self.__block_forward(*args, **kwargs)
  File "/home/cheng/repos/oneflow/python/oneflow/nn/graph/proxy.py", line 238, in __block_forward
    result = unbound_forward_of_module_instance(self, *args, **kwargs)
  File "/home/cheng/pyvenv/oneflow/lib/python3.10/site-packages/diffusers/models/attention.py", line 329, in forward
    attn_output = self.attn1(
  File "/home/cheng/repos/oneflow/python/oneflow/nn/graph/proxy.py", line 188, in __call__
    result = self.__block_forward(*args, **kwargs)
  File "/home/cheng/repos/oneflow/python/oneflow/nn/graph/proxy.py", line 238, in __block_forward
    result = unbound_forward_of_module_instance(self, *args, **kwargs)
  File "/home/cheng/repos/onediff/src/infer_compiler_registry/register_diffusers/attention_processor_oflow.py", line 364, in forward
    return self.processor(
  File "/home/cheng/repos/oneflow/python/oneflow/nn/graph/proxy.py", line 188, in __call__
    result = self.__block_forward(*args, **kwargs)
  File "/home/cheng/repos/oneflow/python/oneflow/nn/graph/proxy.py", line 238, in __block_forward
    result = unbound_forward_of_module_instance(self, *args, **kwargs)
  File "/home/cheng/repos/oneflow/python/oneflow/nn/modules/module.py", line 200, in forward
    raise NotImplementedError()
NotImplementedError
```

File: src/onediff/infer_compiler/transform/builtin_transform.py [hunk 1]
```diff
@@ -149,6 +149,10 @@ def torch2oflow(mod, *args, **kwargs):
 
 
 def default_converter(obj, verbose=False, *, proxy_cls=None):
+    # Higher versions of diffusers might use torch.nn.modules.Linear
+    if obj is torch.nn.Linear:
+        return flow.nn.Linear
```
Comment (hjchen2) [inline src/onediff/infer_compiler/transform/builtin_transform.py:154]: 和BatchNorm一样单独写一个torch2oflow.register比较好， https://github.com/siliconflow/onediff/pull/597/files#diff-b70093b97465c7c52352f21391c1e5bbff4d4a9140dfe197bc88e4bfaa8c5fb5R243-R248

File: benchmarks/instant_id.py [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+        with open(attention_processor_path, "r") as f:
+            content = f.read()
+
+        with open(attention_processor_path, "w") as f:
+            content = content.replace("__call__", "forward")
+            f.write(content)
+
+    sys.path.insert(0, args.repo)
+
+    from pipeline_stable_diffusion_xl_instantid import (
```
Comment (doombeaker) [inline benchmarks/instant_id.py:194]: 这个 `pipeline_stable_diffusion_xl_instantid` 在这个 PR 里没有包含，它是可以有 pip 的安装方式了吗（我自己试跑时是直接 copy 了原作者的文件）

File: src/onediff/infer_compiler/transform/builtin_transform.py (hunk 1 above)
Comment (strint) [reply src/onediff/infer_compiler/transform/builtin_transform.py:154]: 这个比较特别，obj 正常是个对象，这里直接是个 type（self.linear_cls = torch.nn.Linear），导致需要打个补丁

File: benchmarks/instant_id.py (hunk 2 above)
Comment (chengzeyi) [reply benchmarks/instant_id.py:194]: 会提示用户去clone一份到本地

**Comment by chengzeyi** [comment]
# RTX 4090

```shell
python3 benchmarks/instant_id.py --repo ~/repos/InstantID/ --face-analysis-root ~/repos/face-analysis-root/ --seed 0 --output-image tmp/instant_id.jpg
```

## Unoptimized

```
Inference time: 5.806s
Iterations per second: 5.453
CUDA Mem after: 20.447GiB
Host Mem after: 7.249GiB

```

instant_id

## Optimized

```
Inference time: 3.221s
Iterations per second: 10.149
CUDA Mem after: 20.155GiB
Host Mem after: 8.458GiB
```

instant_id

**Review by doombeaker — APPROVED** [review]

File: benchmarks/instant_id.py [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+        # Patch the attention processors to make them compatible with us.
+        attention_processor_path = os.path.join(
+            args.repo, "ip_adapter", "attention_processor.py"
+        )
+        with open(attention_processor_path, "r") as f:
+            content = f.read()
+
+        with open(attention_processor_path, "w") as f:
+            content = content.replace("__call__", "forward")
+            f.write(content)
```
[the remark this answers is not in the document]
Comment (chengzeyi) [reply benchmarks/instant_id.py:190]: 先凑活了，主要是我试了下想dynamic patch上去发现oneflow跑还是会有bug😢

File: src/onediff/infer_compiler/transform/builtin_transform.py (hunk 1 above)
Comment (strint) [inline src/onediff/infer_compiler/transform/builtin_transform.py:154]: 这里应该可以做成通用的，即支持 obj 不是对象，就是一个 torch 的 class

**Comment by plienhar** [comment]
Hi @chengzeyi ,

**Comment by chengzeyi** [comment]
> Hi @chengzeyi , Could you provide a set of revisions for oneflow/onediff/onediffx for which compilation of the InstantID pipeline works? I cannot manage to make it work on my script.

You need to use InstantID with `diffusers==0.25.1`.

**Comment by jamesng93** [comment]
HI @chengzeyi , I got this error when run instantid with onediff
`NotImplementedError: Transform failed of <class 'diffusers.models.unets.unet_2d_condition.UNet2DConditionModel'>: Transform failed of <class 'torch.nn.modules.container.ModuleList'>: Transform failed of <class 'diffusers.models.unets.unet_2d_blocks.CrossAttnDownBlock2D'>: Transform failed of <class 'torch.nn.modules.container.ModuleList'>: Transform failed of <class 'diffusers.models.transformers.transformer_2d.Transformer2DModel'>: Transform failed of <class 'torch.nn.modules.container.ModuleList'>: Transform failed of <class 'diffusers.models.attention.BasicTransformerBlock'>: Transform failed of <class 'diffusers.models.attention_processor.Attention'>: Transform failed of <class 'ip_adapter.attention_processor.AttnProcessor2_0'>: super(type, obj): obj must be an instance or subtype of type`

**Pull request merged**